File: Templates/Day_4/DerivedDatatypes/Fortran/derivedtransfer.f90
Program ReductionTest

  use mpi
  use realint_pair

  Implicit None

  type(ri_pair) :: my_values, reduced_values

  Integer :: my_rank, world_size
  Integer(kind=mpi_address_kind) :: start_address, secnd_address

  Integer, parameter :: nb_entries=2
  Integer, dimension(nb_entries) :: bleng, types
  Integer(kind=mpi_address_kind),dimension(nb_entries) :: displ

  Integer :: ri_mpitype, ri_addmax_mpiop

  Integer, dimension(Mpi_Status_size) :: mstatus
  Integer :: merror

  Call MPI_Init(merror)

  Call MPI_Comm_rank(MPI_Comm_world, my_rank, merror)
  Call MPI_Comm_size(MPI_Comm_world, world_size, merror)

  my_values%realpart = my_rank
  my_values%intpart  = my_rank

  print *,"On rank", my_rank," my_values contains", my_values

  
  ! implement a datatype to send my_values from rank 1 to rank 0


  If (my_rank == 1) then
     Call MPI_Send(my_values, 1, ri_mpitype, 0, 0, MPI_Comm_world, merror)
  elseif (my_rank == 0) then
     Call MPI_Recv(reduced_values, 1, ri_mpitype, 1, 0, MPI_comm_world, &
          mstatus, merror)
     print *,"On rank", my_rank," received value: ", reduced_values
  endif

  Call MPI_Finalize(merror)

End Program ReductionTest

File: Templates/Day_4/DerivedDatatypes/C/derivedtransfer.c
#include <stdio.h>
#include <math.h>
#include "mpi.h"
#include "realint_pair.h"


void setup_mpi(MPI_Comm* pcomm, int* psize, int* prank)
{
    MPI_Init(NULL, NULL);
    
    *pcomm = MPI_COMM_WORLD;
    
    MPI_Comm_size(*pcomm, psize);
    MPI_Comm_rank(*pcomm, prank);
    
    return;
}


int main(int argc, char **argv) 
{

  // setting up MPI

  MPI_Comm my_world;
  int my_rank, world_size;

  setup_mpi(&my_world, &world_size, &my_rank);

  ri_pair my_values, received_values, reduced_values;

  my_values.rpart = my_rank;
  my_values.ipart = my_rank;

  printf("On rank %d my values are: (%f, %d)\n", 
	 my_rank, my_values);
 
  // implement a datatype to send my_values from rank 1 to rank 0
  
  if (my_rank == 0)
    {
      MPI_Status stat;
      MPI_Recv(&received_values, 1, ri_mpitype, 1, 1, my_world, &stat);
      printf("On rank 0 we received from 1: (%f, %d)\n", 
	     received_values);
    }
  else if(my_rank == 1)
    {
      MPI_Send(&my_values, 1, ri_mpitype, 0, 1, my_world);
    }

  MPI_Finalize();

  return 0;
}

